job/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -272,6 +292,50 @@ private void copyLogs() {
             }
         }
     }
+
+    private String getLogPrefix(FlowNode node) {
+        BodyInvocationAction bodyInvocationAction = node.getAction(BodyInvocationAction.class);
+
+        if (bodyInvocationAction != null) {
```

Comment:
Right… I wonder if you can terminate the recursion upon encountering a `BlockEndNode`. Thus `getLogPrefix` on the join node created by the end of the `parallel` step would get `null` from the recursion on each of its parents, which is what you want.